src/Database/LSMTree/Normal.hs

```diff
@@ -252,15 +266,23 @@ data VerificationFailure
 
 data SnapshotName
 
--- | Take a snapshot.
+-- | Make the current value of a table durable on-disk by taking a snapshot and
+-- giving the snapshot a name. This is the _only_ mechanism to make a table
+-- durable -- ordinary insert\/delete operations are otherwise not preserved.
+--
+-- Snapshots have names and the table may be opened later using 'open' via that
+-- name. Names are strings and the management of the names is up to the user of
+-- the library.
+--
+-- The names correspond to disk files, which imposes some constraints on length
+-- and what characters can be used.
 --
 -- Snapshotting does not close the table handle.
 --
--- Taking a snapshot should be relatively cheap, but it is not so cheap that one
--- can use one after every operation. It is relatively cheap because of a reason
--- similar to the reason why 'duplicate' is cheap. On-disk and in-memory data is
--- immutable, which means we can write (the relatively small) in-memory parts to
--- disk, and create hard links for existing files.
+-- Taking a snapshot is /relatively/ cheap, but it is not so cheap that one can
+-- use it after every operation. In the implementation, it must at least flush
+-- the write buffer to disk.
+--
```

Comment:
So this is essentially `writeFile`? What if two concurrent `snapshot`s are taken with the same name, is it a race?

Reply:
Yes, that would be a race.